pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/TimePredicateFilterOptimizer.java

```diff
@@ -0,0 +1,330 @@
+
+  @Override
+  public FilterQueryTree optimize(FilterQueryTree filterQueryTree, @Nullable Schema schema) {
+    // Do not rewrite PQL queries because PQL is deprecated
+    return filterQueryTree;
+  }
+
+  @Override
+  public Expression optimize(Expression filterExpression, @Nullable Schema schema) {
+    return optimize(filterExpression);
+  }
+
+  @VisibleForTesting
+  Expression optimize(Expression filterExpression) {
+    Function filterFunction = filterExpression.getFunctionCall();
+    FilterKind filterKind = FilterKind.valueOf(filterFunction.getOperator());
+    List<Expression> operands = filterFunction.getOperands();
+    if (filterKind == FilterKind.AND || filterKind == FilterKind.OR) {
+      operands.replaceAll(this::optimize);
+    } else if (filterKind.isRange() || filterKind == FilterKind.EQUALS) {
+      Expression expression = operands.get(0);
+      if (expression.getType() == ExpressionType.FUNCTION) {
+        Function expressionFunction = expression.getFunctionCall();
+        String functionName = StringUtils.remove(expressionFunction.getOperator(), '_');
+        if (functionName.equalsIgnoreCase(TimeConversionTransformFunction.FUNCTION_NAME)) {
+          optimizeTimeConvert(filterFunction, filterKind);
+        } else if (functionName.equalsIgnoreCase(DateTimeConversionTransformFunction.FUNCTION_NAME)) {
+          optimizeDateTimeConvert(filterFunction, filterKind);
+        }
+      }
+    }
+    return filterExpression;
+  }
+
+  /**
+   * Helper method to optimize TIME_CONVERT function with range/equality predicate to directly apply the predicate to
+   * the inner expression. Changes are applied in-place of the filter function.
+   */
+  private void optimizeTimeConvert(Function filterFunction, FilterKind filterKind) {
+    List<Expression> filterOperands = filterFunction.getOperands();
+    List<Expression> timeConvertOperands = filterOperands.get(0).getFunctionCall().getOperands();
+    Preconditions.checkArgument(timeConvertOperands.size() == 3,
+        "Exactly 3 arguments are required for TIME_CONVERT transform function");
+    Preconditions
+        .checkArgument(isStringLiteral(timeConvertOperands.get(1)) && isStringLiteral(timeConvertOperands.get(2)),
+            "The 2nd and 3rd argument for TIME_CONVERT transform function must be string literal");
+
+    try {
+      TimeUnit inputTimeUnit = TimeUnit.valueOf(timeConvertOperands.get(1).getLiteral().getStringValue().toUpperCase());
+      TimeUnit outputTimeUnit =
+          TimeUnit.valueOf(timeConvertOperands.get(2).getLiteral().getStringValue().toUpperCase());
+
+      // Step 1: Convert output range to millis range
+      Long lowerMillis = null;
+      Long upperMillis = null;
+      switch (filterKind) {
+        case GREATER_THAN: {
+          // millisToFormat(millis) > n
+          // -> millisToFormat(millis) >= n + 1
+          // -> millis >= formatToMillis(n + 1)
```

Comment:
Not sure why you need to change the input value here by doing +1. Seems like all that needs to be done here is to properly set lowerInclusive boolean value?

Reply:
We have to do `+1` here, or it won't be correct. E.g. `millisToSeconds(millis) > 0` is not equivalent to `millis > 0` because of the flooring of the value. Will add the example into the comment